### PR TITLE
Feature: move topic to another messageboard

### LINKED ISSUE #507: Feature: move topic to another messageboard
Allow users with permissions to edit a topic to re-assign the topic to any messageboard they can post in.

### PR DESCRIPTION
**Pull request by glebm**
Allow users with permissions to edit a topic to re-assign the topic to
any messageboard they can post in.

Fixes #507

### CODE REVIEW

File: app/forms/thredded/topic_form.rb [hunk 1]
```diff
@@ -22,12 +22,8 @@ def self.model_name
       Thredded::Topic.model_name
     end
 
-    def categories
-      topic.messageboard.categories
-    end
-
     def category_options
-      categories.map { |cat| [cat.name, cat.id] }
+      messageboard.categories.map { |cat| [cat.name, cat.id] }
```
Comment (houndci-staging) [inline app/forms/thredded/topic_form.rb]: FeatureEnvy: Thredded::TopicForm#category_options refers to 'cat' more than self (maybe move it to another class?)

File: app/forms/thredded/edit_topic_form.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+module Thredded
+  class EditTopicForm
```
Comment (houndci-staging) [inline app/forms/thredded/edit_topic_form.rb]: IrresponsibleModule: Thredded::EditTopicForm has no descriptive comment

File: app/forms/thredded/edit_topic_form.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    def self.model_name
+      Thredded::Topic.model_name
+    end
+
+    def category_options
+      @topic.messageboard.categories.map { |cat| [cat.name, cat.id] }
+    end
+
+    def messageboard_options
+      @user.thredded_can_write_messageboards.map { |messageboard| [messageboard.name, messageboard.id] }
```
Comment (houndci-staging) [inline app/forms/thredded/edit_topic_form.rb]: FeatureEnvy: Thredded::EditTopicForm#messageboard_options refers to 'messageboard' more than self (maybe move it to another class?)

File: app/forms/thredded/edit_topic_form.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      @user = user
+      @topic = topic
+    end
+
+    def self.model_name
+      Thredded::Topic.model_name
+    end
+
+    def category_options
+      @topic.messageboard.categories.map { |cat| [cat.name, cat.id] }
```
Comment (houndci-staging) [inline app/forms/thredded/edit_topic_form.rb]: FeatureEnvy: Thredded::EditTopicForm#category_options refers to 'cat' more than self (maybe move it to another class?)

File: app/controllers/thredded/topics_controller.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 module Thredded
-  class TopicsController < Thredded::ApplicationController
+  class TopicsController < Thredded::ApplicationController # rubocop:disable Metrics/ClassLength
```
Comment (houndci-staging) [inline app/controllers/thredded/topics_controller.rb]: InstanceVariableAssumption: Thredded::TopicsController assumes too much for instance variable '@category'<br>InstanceVariableAssumption: Thredded::TopicsController assumes too much for instance variable '@edit_topic'<br>InstanceVariableAssumption: Thredded::TopicsController assumes too much for instance variable '@new_topic'<br>InstanceVariableAssumption: Thredded::TopicsController assumes too much for instance variable '@query'<br>IrresponsibleModule: Thredded::TopicsController has no descriptive comment<br>TooManyInstanceVariables: Thredded::TopicsController has at least 7 instance variables<br>TooManyMethods: Thredded::TopicsController has at least 16 methods

File: app/controllers/thredded/topics_controller.rb [hunk 6]
```diff
@@ -92,12 +92,23 @@ def create
...
       authorize topic, :update?
-      if topic.update(topic_params.merge(last_user_id: thredded_current_user.id))
-        redirect_to messageboard_topic_url(messageboard, topic),
+      if topic.messageboard_id_changed?
+        # Work around the association not being reset.
+        # TODO: report issue to Rails. Looks like a regression of:
+        # https://rails.lighthouseapp.com/projects/8994/tickets/2989
+        topic.messageboard = Thredded::Messageboard.find(topic.messageboard_id)
+
+        authorize topic.messageboard, :post?
```
Comment (houndci-staging) [inline app/controllers/thredded/topics_controller.rb]: DuplicateMethodCall: Thredded::TopicsController#update calls 'topic.messageboard' 2 times

File: app/forms/thredded/topic_form.rb (hunk 1 above)
Comment (jayroh) [inline app/forms/thredded/topic_form.rb]: sshhhh hound ssshhh

**Comment by jayroh** [comment]
👍 from me

(a nice derivative feature, if it's not in the issues list yet - being able to split out a topic from one post onwards to a new topic.)

**Pull request merged**